metabase/metabase
# Disable auto-apply filters via a toaster

M2 of an epic #29267

### Description

Show a toast to users that open a dashboard and it loads slower than 15 seconds, and has parameter value set. Users can turn off auto-apply filters option via the action button in the toast.


```mermaid
stateDiagram-v2
never: Never show toast
ready: Ready for toast
loading: Loading cards
timedOut: Timed out
shown: Toast visible

never --> ready: ready condition
ready --> never: never condition

ready --> loading: load()

loading --> timedOut: timed out (15s)
loading --> loading: load()
loading --> never: never condition
loading --> ready: loaded

timedOut --> loading: load()
timedOut --> shown: loaded
timedOut --> never: never condition

shown --> never: never condition \n(turn-off auto-apply filters)
shown --> ready: dismiss toast

note left of never
  Never condition.
  When`auto_apply_filters: false`
  OR has no `parameterValues`
end note

note left of ready
  Ready condition.
  When`auto_apply_filters: true`
  AND has `parameterValues`
end note
```

### How to verify

If you can make your dashboard load so slow that it will show the slow toast, just make sure your dashboard has at least one parameter and at least one of the parameters has the value set. Then after the slow toast disappear, the dashboard will load together with showing the auto-apply filter toasts.

But it's a lot easier to mock all the timeout/delay behavior using a few lines of code change
<details>
<summary>Steps to do that</summary>

1. Go to [`frontend/src/metabase/dashboard/constants.js`](https://github.com/metabase/metabase/blob/03a5c61fe9e10ff18a3ea31164e413f6a527c7c4/frontend/src/metabase/dashboard/constants.js#L10) and edit https://github.com/metabase/metabase/blob/03a5c61fe9e10ff18a3ea31164e413f6a527c7c4/frontend/src/metabase/dashboard/constants.js#L10
    to `2 * 1000`. That will make the slow toast appear after 2 seconds. You can adjust the number as you want, but I set it to a relatively low value so I feel I need to wait but not for too long.
2. go to `frontend/src/metabase/dashboard/actions/data-fetching.js` and paste these 2 functions at the bottom of the file
    ```ts
    function randomBetween(min, max) {
      return Math.random() * (max - min) + min;
    }

    function delay(timeout) {
      return new Promise(resolve => {
        setTimeout(resolve, timeout);
      });
    }
    ```
    Then add this expression at [line 362](https://github.com/metabase/metabase/blob/03a5c61fe9e10ff18a3ea31164e413f6a527c7c4/frontend/src/metabase/dashboard/actions/data-fetching.js#L362) `await delay(randomBetween(2000, 4000));`, that part should look something like this.
    ```ts
    await delay(randomBetween(2000, 4000));
    setFetchCardDataCancel(card.id, dashcard.id, null);
    clearTimeout(slowCardTimer);
    ```
     You can easily tweak this number to anything you like. This will mock the duration it takes to load each card on your dashboard. In this case, I random that a card would load between 2 and 4 seconds. You could even delay this longer to see the slow toast longer since it will disappear once all cards are loaded. And after that, you should be able to see the auto-apply filters notification toast (given all conditions are met, according to the spec i.e. the slow toast is shown + auto-apply filter is true, and there is at least a single parameter with values)
</details>

### Demo
I mock the API delay, and change the delay value so that we don't need to wait for the slow toast for too long.

https://user-images.githubusercontent.com/1937582/234380975-53e2825f-8c99-4086-994b-709094707b45.mov


### Checklist



- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30371)
